storage: fix replica removal when learners not active

tbg opened the pull request:
I introduced a bug in #39640 which would fail removal of replicas
whenever preemptive snapshots were used, since we'd accidentally
send a preemptive snapshot which would end up with a "node already
in descriptor" error.

See:

https://github.com/cockroachdb/cockroach/pull/39640#pullrequestreview-275532068

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39697)
<!-- Reviewable:end -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39697)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/server/version_cluster_test.go, line 403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-LmMRYfG6qYGdXL9xB6q:-LmMRYfG6qYGdXL9xB6r:b46huod) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ea1cdd807b60ac973fb9b6a77d29db0f53007d/pkg/server/version_cluster_test.go#L403)):*
> ```Go
> 		k := tc.ScratchRange(t)
> 		tc.AddReplicasOrFatal(t, k, tc.Target(2))
> 		log.Infof(ctx, "TBG====")
> ```

intentional?
___
*[pkg/storage/replica_command.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-LmMRdjHD-V9-f_O--Tv:-LmMRdjHD-V9-f_O--Tw:bwwtqrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ea1cdd807b60ac973fb9b6a77d29db0f53007d/pkg/storage/replica_command.go#L981)):*
> ```Go
> 			target := chgs[0].Target
> 			return r.addReplicaLegacyPreemptiveSnapshot(ctx, target, desc, priority, reason, details)
> 		}
> ```

this `!useLearners` branch works now by falling through to the learners codepath, but happening to not do any of the learner stuff because there aren't any replica additions. this is pretty subtle. i'd much prefer to have something more direct here (like an else that calls `finalizeChangeReplicas`, though the name of that is a little confusing in this context.)


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
PTAL (ignore the first commit, it's https://github.com/cockroachdb/cockroach/pull/39690)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-:-LmObKf0By2fjZC9y88v:b-llwxif)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/server/version_cluster_test.go, line 403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-LmMRYfG6qYGdXL9xB6q:-LmOX3861E-zq-5awmrb:by1gwsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ea1cdd807b60ac973fb9b6a77d29db0f53007d/pkg/server/version_cluster_test.go#L403)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

intentional?
</blockquote></details>

:see_no_evil: just checking that you're paying attention
Removed.
___
*[pkg/storage/replica_command.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-LmMRdjHD-V9-f_O--Tv:-LmOY99u1OuYPoJBD5Ip:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ea1cdd807b60ac973fb9b6a77d29db0f53007d/pkg/storage/replica_command.go#L981)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this `!useLearners` branch works now by falling through to the learners codepath, but happening to not do any of the learner stuff because there aren't any replica additions. this is pretty subtle. i'd much prefer to have something more direct here (like an else that calls `finalizeChangeReplicas`, though the name of that is a little confusing in this context.)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-:-LmQ5PiF3psQ_PyYZTdH:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/storage/replica_command.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39697#-LmMRdjHD-V9-f_O--Tv:-LmQ5ML1B9t602PL6TCV:b-tn44u1) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ea1cdd807b60ac973fb9b6a77d29db0f53007d/pkg/storage/replica_command.go#L981)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

much better thank you


<!-- Sent from Reviewable.io -->

tbg committed:
storage: fix replica removal when learners not active

I introduced a bug in #39640 which would fail removal of replicas
whenever preemptive snapshots were used, since we'd accidentally
send a preemptive snapshot which would end up with a "node already
in descriptor" error.

See:

https://github.com/cockroachdb/cockroach/pull/39640#pullrequestreview-275532068

Release note: None

tbg commented:
TFTR! Fixed the lint error so this should be green now

bors r=danhhz
